go/engine/selfprovision.go

```diff
@@ -0,0 +1,260 @@
+package engine
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+)
+
+type SelfProvisionEngine struct {
+	libkb.Contextified
+	DeviceName     string
+	result         error
+	lks            *libkb.LKSec
+	User           *libkb.User
+	perUserKeyring *libkb.PerUserKeyring
+}
+
+// If a device is cloned, we can provision a new device from the current device
+// to get out of the cloned state.
+func NewSelfProvisionEngine(g *libkb.GlobalContext, deviceName string) *SelfProvisionEngine {
+	return &SelfProvisionEngine{
+		Contextified: libkb.NewContextified(g),
+		DeviceName:   deviceName,
+	}
+}
+
+func (e *SelfProvisionEngine) Name() string {
+	return "SelfProvision"
+}
+
+func (e *SelfProvisionEngine) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+func (e *SelfProvisionEngine) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.ProvisionUIKind,
+		libkb.LogUIKind,
+		libkb.SecretUIKind,
+		libkb.LoginUIKind,
+	}
+}
+
+func (e *SelfProvisionEngine) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{
+		&loginLoadUser{},
+	}
+}
+
+func (e *SelfProvisionEngine) Result() error {
+	return e.result
+}
+
+func (e *SelfProvisionEngine) Run(m libkb.MetaContext) (err error) {
+	m.G().LocalSigchainGuard().Set(m.Ctx(), "SelfProvisionEngine")
+	defer m.G().LocalSigchainGuard().Clear(m.Ctx(), "SelfProvisionEngine")
+	defer m.CTrace("SelfProvisionEngine#Run", func() error { return err })()
+
+	if d, err := libkb.GetDeviceCloneState(m); err != nil {
+		return err
+	} else if !d.IsClone() {
+		return fmt.Errorf("to self provision, you must be a cloned device")
+	}
+
+	// transaction around config file
+	tx, err := e.G().Env.GetConfigWriter().BeginTransaction()
+	if err != nil {
+		return err
+	}
+
+	uv, _ := e.G().ActiveDevice.GetUsernameAndUserVersionIfValid(m)
+	// Pass the UV here so the passphrase stream is cached on the provisional
+	// login context
+	m = m.WithNewProvisionalLoginContextForUserVersionAndUsername(uv, e.G().Env.GetUsername())
+
+	// From this point on, if there's an error, we abort the transaction.
+	defer func() {
+		if tx != nil {
+			tx.Abort()
+		}
+		if err == nil {
+			m = m.CommitProvisionalLogin()
```

Comment:
Isn't this a noop?

Reply:
the assignment is but the call caches the passphrase stream onto the active device so we still want it. fixed and added a comment to clarify